apps/prairielearn/src/ee/pages/institutionAdminLti13/institutionAdminLti13.ts

```diff
@@ -0,0 +1,192 @@
+        resLocals: res.locals,
+        platform_defaults,
+      }),
+    );
+  }),
+);
+
+router.post(
+  '/:lti13_instance_id?',
+  asyncHandler(async (req, res) => {
+    if (req.body.__action === 'add_key') {
+      const keystoreJson = await queryAsync(sql.select_keystore, {
+        lti13_instance_id: req.params.lti13_instance_id,
+        institution_id: req.params.institution_id,
+      });
+      const keystore = await jose.JWK.asKeyStore(keystoreJson?.rows[0]?.keystore || []);
+
+      const kid = new Date().toUTCString();
+      // RSA256 minimum keysize of 2048 bits
+      await keystore.generate('RSA', 2048, {
+        alg: 'RS256',
+        use: 'sig',
+        kid: kid,
+      });
+
+      await queryAsync(sql.update_keystore, {
+        lti13_instance_id: req.params.lti13_instance_id,
+        institution_id: req.params.institution_id,
+        // true to include private keys
+        keystore: keystore.toJSON(true),
+      });
+      flash('success', `Key ${kid} added.`);
+    } else if (req.body.__action === 'delete_keys') {
+      await queryAsync(sql.update_keystore, {
+        lti13_instance_id: req.params.lti13_instance_id,
+        institution_id: req.params.institution_id,
+        keystore: null,
+      });
+      flash('success', `All keys deleted.`);
+    } else if (req.body.__action === 'delete_key') {
+      const keystoreJson = await queryAsync(sql.select_keystore, {
+        lti13_instance_id: req.params.lti13_instance_id,
+        institution_id: req.params.institution_id,
+      });
+      const keystore = await jose.JWK.asKeyStore(keystoreJson?.rows[0]?.keystore || []);
+
+      const key = keystore.get(req.body.kid);
+
+      // Validate the key before removal because keystore.get() returns the first key
+      if (req.body.kid === key.kid) {
+        keystore.remove(key);
+
+        await queryAsync(sql.update_keystore, {
+          lti13_instance_id: req.params.lti13_instance_id,
+          institution_id: req.params.institution_id,
+          // true to include private keys
+          keystore: keystore.toJSON(true),
+        });
+      } else {
+        throw error.make(500, 'error removing key', {
+          locals: res.locals,
+          body: req.body,
+        });
+      }
+      flash('success', `Key ${key.kid} deleted.`);
+    } else if (req.body.__action === 'update_platform') {
+      const url = getCanonicalHost(req);
+
+      const client_params = {
+        client_id: req.body.client_id || null,
+        redirect_uris: [`${url}/pl/lti13_instance/${req.params.lti13_instance_id}/auth/callback`],
+        token_endpoint_auth_method: 'private_key_jwt',
+        token_endpoint_auth_signing_alg: 'RS256',
+      };
+
+      await queryAsync(sql.update_platform, {
+        lti13_instance_id: req.params.lti13_instance_id,
+        institution_id: req.params.institution_id,
+        issuer_params: req.body.issuer_params,
+        platform: req.body.platform,
+        client_params,
```

Comment:
Both `issuer_params` and `client_params` go directly into [openid-client's Client() and Issuer() calls](https://github.com/panva/node-openid-client/blob/main/docs/README.md#new-clientmetadata-jwks-options). My thinking was, not knowing what all of those settings might need to be for different LMS configurations, and because PL doesn't really interact with it (other than to hand it off to the library), we can store it independently and just pass it into the functions. As we learn more about what's needed for the LTI 1.3 OAuth config, we can move things more into columns: client_params is a good example of how that evolved as we're storing the `client_id` as an independent variable/column and load it into the `client_params` on save.

Reply:
I'm fine to store these as JSON blobs for now.